src/coalgebras/itreeTauScript.sml

```diff
@@ -1564,6 +1614,1047 @@ Proof
+                                  !l. itree_wbisim (k l) (k' l))
+Proof
+  iff_tac
+  >- (strip_tac
+      \\ reverse $ Cases_on ‘?t''. strip_tau t t''’ \\ gvs[]
+      >- (drule strip_tau_spin
+          \\ metis_tac[wbisim_spin_eq, itree_wbisim_sym]
+         )
+      \\ Cases_on ‘t''’ \\ gvs[]
+      >- (drule_all itree_wbisim_strip_tau_Ret
+          \\ metis_tac[]
+         )
+      \\ drule_all itree_wbisim_strip_tau_Vis
+      \\ metis_tac[]
+     )
+  \\ rpt strip_tac
+  >- rw[wbisim_spin_eq]
+  \\ metis_tac[itree_wbisim_rules]
+QED
+
+Theorem after_taus_itree_strong_bisim_spin_spin:
+  after_taus ($=) t t' ==> t' = spin ==> t = spin
+Proof
+  qid_spec_tac ‘t'’
+  \\ qid_spec_tac ‘t’
+  \\ ho_match_mp_tac after_taus_strongind
+  \\ rw[spin, GSYM spin]
+  \\ ‘Tau t' = Tau spin’ by metis_tac[spin]
+  \\ gvs[]
+QED
+
+Theorem after_taus_itree_strong_bisim_strip_tau:
+  after_taus ($=) t t' <=> (?t''. strip_tau t t'' /\ strip_tau t' t'') \/ (t = spin /\ t' = spin)
+Proof
+  iff_tac
+  >- (qid_spec_tac ‘t'’
+      \\ qid_spec_tac ‘t’
+      \\ ho_match_mp_tac after_taus_strongind
+      \\ rw[spin, GSYM spin]
+      \\ metis_tac[strip_tau_spin, spin]
+     )
+  \\ reverse $ rw[]
+  >- (irule after_taus_rel \\ rw[]
+     )
+  \\ imp_res_tac strip_tau_FUNPOW
+  \\ rw[]
+  \\ irule after_taus_FUNPOW_TauL
+  \\ irule after_taus_FUNPOW_TauR
+  \\ irule after_taus_rel \\ rw[]
+QED
+
+Theorem after_taus_itree_wbisim_spin_spin:
+  after_taus itree_wbisim t t' ==> t' = spin ==> t = spin
+Proof
+  qid_spec_tac ‘t'’
+  \\ qid_spec_tac ‘t’
+  \\ ho_match_mp_tac after_taus_strongind
+  \\ gvs[spin, GSYM spin, wbisim_spin_eq]
+  \\ rpt strip_tac
+  \\ ‘Tau t' = Tau spin’ by metis_tac[spin]
+  \\ gvs[]
+QED
+
+Theorem after_taus_itree_wbisim_strip_tau:
+  after_taus itree_wbisim t t' <=>
+  (?t'' t'''. strip_tau t t'' /\ strip_tau t' t''' /\ itree_wbisim t'' t''') \/
+  (t = spin /\ t' = spin)
+Proof
+  iff_tac
+  >- (qid_spec_tac ‘t'’
+      \\ qid_spec_tac ‘t’
+      \\ ho_match_mp_tac after_taus_strongind
+      \\ rw[spin, GSYM spin]
+      >- (reverse $ Cases_on ‘?x. strip_tau t x’ \\ gvs[]
+          >- (dxrule strip_tau_spin
+              \\ rw[]
+              \\ irule $ iffLR wbisim_spin_eq
+              \\ rw[itree_wbisim_sym]
+             )
+          \\ Cases_on ‘x’ \\ gvs[]
+          >- (qspecl_then [‘t’, ‘t'’, ‘x'’] assume_tac itree_wbisim_strip_tau_Ret
+              \\ metis_tac[itree_wbisim_refl]
+             )
+          \\ qspecl_then [‘t’, ‘t'’, ‘a’, ‘g’] assume_tac itree_wbisim_strip_tau_Vis
+          \\ metis_tac[itree_wbisim_vis]
+         )
+      \\ metis_tac[]
+     )
+  \\ reverse $ rw[]
+  >- (irule after_taus_rel \\ rw[itree_wbisim_refl]
+     )
+  \\ imp_res_tac strip_tau_FUNPOW
+  \\ rw[]
+  \\ irule after_taus_FUNPOW_TauL
+  \\ irule after_taus_FUNPOW_TauR
+  \\ irule after_taus_rel \\ rw[]
+QED
+
+(* return or eventually reach another tree *)
+CoInductive ret_or_reach:
```

Comment:
I find it a bit odd that `ret_or_reach` uses `strip_tau` style in the first two clauses, but `FUNPOW` style in the third. the two styles are obviously equivalent and interconvertible, but it would be nice to be consistent about it within the one definition.